mimic/rest/auth_api.py

```diff
@@ -2,38 +2,107 @@
+from mimic.model.behaviors import BehaviorRegistry, EventDescription
+from mimic.model.identity import (
+    APIKeyCredentials,
+    ImpersonationCredentials,
+    PasswordCredentials,
+    TokenCredentials)
 from mimic.rest.mimicapp import MimicApp
-from mimic.canned_responses.auth import format_timestamp
-from mimic.util.helper import invalid_resource
 from mimic.session import NonMatchingTenantError
-from mimic.canned_responses.mimic_presets import get_presets
+from mimic.util.helper import invalid_resource
 
 Request.defaultContentType = 'application/json'
 
 
-class AuthApi(object):
+authentication = EventDescription()
+"""
+Event refers to authenticating against Identity using a username/password,
+username/api-key, token, or getting an impersonation token.
+"""
+
+
+@authentication.declare_default_behavior
+def default_authentication_behavior(core, http_request, credentials):
+    """
+    Default behavior in response to a server creation.
+
+    :param core: An instance of :class:`mimic.core.MimicCore`
+    :param http_request: A twisted http request/response object
+    :param credentials: An `mimic.model.identity.ICredentials` provider
+    """
+    if type(credentials) in (
+            APIKeyCredentials, PasswordCredentials, TokenCredentials):
+        try:
+            session = credentials.get_session(core.sessions)
+        except NonMatchingTenantError as e:
+            http_request.setResponseCode(401)
+            if type(credentials) == TokenCredentials:
+                message = ("Token doesn't belong to Tenant with Id/Name: "
+                           "'{0}'".format(e.desired_tenant))
+            else:
+                message = ("Tenant with Name/Id: '{0}' is not valid for "
+                           "User '{1}' (id: '{2}')".format(
+                               e.desired_tenant,
+                               e.session.username,
+                               e.session.user_id))
+
+            return json.dumps({
+                "unauthorized": {
+                    "code": 401,
+                    "message": message
+                }
+            })
+        else:
+            http_request.setResponseCode(200)
+            prefix_map = {
+                # map of entry to URI prefix for that entry
+            }
+
+            def lookup(entry):
+                return prefix_map[entry]
+            result = get_token(
+                session.tenant_id,
+                entry_generator=lambda tenant_id:
+                list(core.entries_for_tenant(
+                     session.tenant_id, prefix_map,
+                     base_uri_from_request(http_request))),
+                prefix_for_endpoint=lookup,
+                response_token=session.token,
+                response_user_id=session.user_id,
+                response_user_name=session.username,
+            )
+            return json.dumps(result)
+    else:
+        session = credentials.get_session(core.sessions)
```

Comment:
Reviewer: should impersonation be an authentication behavior?
Arguments for:

1.  It is, in a sense, authenticating - it is doing some session creation.
1.  It could use the same behavior criterion as the other authentication creds (username, token)

Arguments against:

1.  It's technically the "auth token" not a token, which comes from the headers and not the JSON blob.